service/datasync/exec_data_sync.go

```diff
@@ -0,0 +1,194 @@
+	e.queue.PushFront(blockID)
+
+	e.log.Debug().Hex("block", blockID[:]).Msg("execution record queued for download")
+}
+
+// Next returns the next available block data. It returns an ErrUnavailable if no block
+// data is available at the moment.
+func (e *ExecDataSync) Next() (*entities.BlockExecutionData, error) {
+
+	// If we are not polling already, we want to start polling in the
+	// background. This will try to fill the buffer up until its limit is
+	// reached. It basically means that the cloud streamer will always be
+	// downloading if something is available and the execution tracker is asking
+	// for the next record.
+	go e.poll()
+
+	// If we have nothing in the buffer, we can return the unavailable error,
+	// which will cause the mapper logic to go into a wait state and retry a bit
+	// later.
+	if e.buffer.Len() == 0 {
+		e.log.Debug().Msg("buffer empty, no execution record available")
+		return nil, dps.ErrUnavailable
+	}
+
+	// If we have a record in the buffer, we will just return it. The buffer is
+	// concurrency safe, so there is no problem with popping from the back while
+	// the poll is pushing new items in the front.
+	record := e.buffer.PopBack()
+	return record.(*entities.BlockExecutionData), nil
+}
+
+func (e *ExecDataSync) poll() {
+
+	// We only call `Next()` sequentially, so there is no need to guard it from
+	// concurrent access. However, when the buffer is not empty, we might still
+	// be polling for new data in the background when the next call happens. We
+	// thus need to ensure that only one poll is executed at the same time. We
+	// do this with a simple flag that is set atomically to work like a
+	// `TryLock()` on a mutex, which is unfortunately not available in Go, see:
+	// https://github.com/golang/go/issues/6123
+	if !atomic.CompareAndSwapUint32(&e.busy, 0, 1) {
+		return
+	}
+	defer atomic.StoreUint32(&e.busy, 0)
+
+	// At this point, we try to pull new files from the cloud.
+	err := e.getExecData()
+	if errors.Is(err, storage.ErrObjectNotExist) {
+		e.log.Debug().Msg("next execution record not available, download stopped")
+		return
+	}
+	if err != nil {
+		e.log.Error().Err(err).Msg("could not download execution records")
+		return
+	}
+}
+
+func (e *ExecDataSync) getExecData() error {
+
+	for {
+
+		// We only want to retrieve and process execData blocks until the buffer is full.
+		if uint(e.buffer.Len()) >= e.limit {
+			e.log.Debug().Uint("limit", e.limit).Msg("buffer full, stopping execution record pull")
+			return nil
+		}
+
+		// We only want to retrieve and process files for blocks that have already
+		// been finalized, in the order that they have been finalized. This
+		// causes some latency, as we don't download until after a block is
+		// finalized, even if the data is available before. However, it seems to
+		// be the only way to make sure trie updates are delivered to the mapper
+		// in the right order without changing the way uploads work.
```

Comment:
this comment doesn't really apply to execution sync data. The AN will only download data for sealed blocks, so the blocks are guaranteed to be sealed here.